pkg/steps/release/source.go

```diff
@@ -0,0 +1,151 @@
+package release
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/release"
+	"github.com/openshift/ci-tools/pkg/release/candidate"
+	"github.com/openshift/ci-tools/pkg/release/official"
+	"github.com/openshift/ci-tools/pkg/release/prerelease"
+	"github.com/openshift/ci-tools/pkg/results"
+	"github.com/openshift/ci-tools/pkg/steps/utils"
+)
+
+// ReleaseSource holds information to resolve a release pull-spec
+// This is used to defer resolution in some cases until after a release is
+// known to be required, based on the pruned step graph.  This structure holds
+// the required configuration and clients until then.
+type ReleaseSource interface {
+	// Input provides the value for the `Inputs` method of steps
+	// The return value may be empty if the pull-spec should not contribute to
+	// the input.
+	Input(ctx context.Context) (string, error)
+	// PullSpec resolves the release pull-spec (if necessary) and returns it
+	PullSpec(ctx context.Context) (string, error)
+}
+
+// NewReleaseSourceFromPullSpec creates a fixed, pre-computed source
+func NewReleaseSourceFromPullSpec(s string) ReleaseSource {
+	ret := fixedReleaseSource(s)
+	return &ret
+}
+
+// NewReleaseSourceFromConfig uses the pull-spec of a published release payload
+func NewReleaseSourceFromConfig(
+	config *api.ReleaseConfiguration,
+	httpClient release.HTTPClient,
+) ReleaseSource {
+	return &configurationReleaseSource{
+		config: config,
+		client: httpClient,
+	}
+}
+
+// NewReleaseSourceFromClusterClaim determines the pull-spec for a cluster pool
+func NewReleaseSourceFromClusterClaim(
+	name string,
+	claim *api.ClusterClaim,
+	hiveClient ctrlruntimeclient.WithWatch,
+) ReleaseSource {
+	return &clusterClaimReleaseSource{
+		testName: name,
+		claim:    claim,
+		client:   hiveClient,
+	}
+}
+
+type fixedReleaseSource string
+
+func (s fixedReleaseSource) PullSpec(ctx context.Context) (string, error) {
+	return string(s), nil
+}
+
+func (s fixedReleaseSource) Input(ctx context.Context) (string, error) {
+	return s.PullSpec(ctx)
+}
+
+type configurationReleaseSource struct {
+	pullSpec string
+	config   *api.ReleaseConfiguration
+	client   release.HTTPClient
+}
+
+func (s configurationReleaseSource) PullSpec(
+	ctx context.Context,
+) (string, error) {
+	if s.pullSpec == "" {
+		if err := s.resolvePullSpec(); err != nil {
+			return "", err
+		}
+	}
+	return s.pullSpec, nil
+}
+
+func (s configurationReleaseSource) Input(ctx context.Context) (string, error) {
+	return s.PullSpec(ctx)
+}
+
+func (s *configurationReleaseSource) resolvePullSpec() (err error) {
```

Comment:
The fact the we have left `s.config.Integration` out makes me sleep bad at night... but I understand that we would need a totally different abstraction to include this whole [piece of code](https://github.com/openshift/ci-tools/blob/56fd6a6664337b8d989ea5940387abc7f08f5bfe/pkg/defaults/defaults.go#L207-L220)

Reply:
Integration streams are a completely different case and not handled at all in the same manner as all the other cases.  They are assembled from tags in an `ImageStream`, not imported as a single image.